lib/react_client.dart

```diff
@@ -36,19 +36,92 @@ newJsMap(Map map) {
 }
 
 /**
- * Type of [children] must be child or list of childs, when child is JsObject or String
+ * Type of [children] must be child or list of children, when child is JsObject or String
  */
 typedef JsObject ReactComponentFactory(Map props, [dynamic children]);
 typedef Component ComponentFactory();
 
 class ReactComponentFactoryProxy implements Function {
-  final ReactComponentFactory _call;
   final JsFunction reactComponentFactory;
-  ReactComponentFactoryProxy(this.reactComponentFactory, this._call);
+  ReactComponentFactoryProxy(this.reactComponentFactory);
 
   JsObject call(Map props, [dynamic children]) {
-    return this._call(props, children);
+    List<dynamic> reactParams = [_generateExtendedJSProps(props, children)];
+    reactParams.add(children);
+    return reactComponentFactory.apply(reactParams);
   }
+
+  dynamic noSuchMethod(Invocation invocation) {
+    if ((invocation.memberName == #call) && (invocation.isMethod)) {
+      Map props = invocation.positionalArguments[0];
+      List<dynamic> children = invocation.positionalArguments.sublist(1);
+      List<dynamic> reactParams = [_generateExtendedJSProps(props, children)];
+      reactParams.addAll(children);
+      return reactComponentFactory.apply(reactParams);
+    }
+    return super.noSuchMethod(invocation);
+  }
+
+  JsObject _generateExtendedJSProps(Map props, dynamic children) {
+
+    Map extendedProps = new Map.from(props);
+    extendedProps['children'] = children;
```

Comment:
_For reference: These changes make `props.children` an opaque data structure, much like [in React JS](https://facebook.github.io/react/tips/children-props-type.html)._

---

Well, having `props['children']` be a List or single item without `React.Children` helpers is doable, just a little more tedious (and requires assumptions in the implementation that don't really leave the representation opaque, but that's a different topic).

##### Iterating over children
These use-case is probably a bit more rare, but the `React.Children.map`/`React.Children.forEach` utility functions  would be easy to implement and make the process a bit less tedious:

```dart
// Before variadic children changes
render() {
  return div({},
      props['children'].map(renderChild)
  );
}
```

```dart
// After variadic children changes
render() {
  dynamic children = props['children'];
  List childrenList = (children is List) ? children : [children];

  return div({},
      childrenList.map(renderChild)
  );
}
```

```dart
// After variadic children changes, with React.children utils
render() {
  return div({},
      react_children.map(props['children'], renderChild)
  );
}
```


##### Counting children
Especially not the end of the world to not have a utility for this, but it'd also be easy to implement:

```dart
// Before variadic children changes
render() {
  List children = props['children'];
  return div({
    'role': children.length > 1 ? 'group' : null
  }, children);
}
```

```dart
// After variadic children changes
render() {
  dynamic children = props['children'];
  return div({
    'role': children is List && children.length > 1 ? 'group' : null
  }, children);
}
```

---

For me, the main concern is making children not always a `List`, given that `react-dart` really only supports flat lists of children (see https://github.com/cleandart/react-dart/issues/60).

Only being able to specify flat lists of children in react-dart makes it simple to concatenate List children, but tedious to do so with children that aren't always Lists. I see this being a pain to refactor.

Before this PR's current changes, adding new rendered items as siblings to children was easy:
```dart
// Assuming props['children'] is always a List (react <0.8)
render() {
  var augmentedChildren = []
    ..add(react.span({}, 'Sibling to children'))
    ..addAll(props['children']);

  return div({}, augmentedChildren);
}
```

After the changes, it becomes a bit more tedious.
```dart
// Assuming props['children'] can be a single child or a List.
render() {
  var augmentedChildren = []
    ..add(react.span({}, 'Sibling to children'));

  dynamic children = props['children'];
  if (children is List) {
    augmentedChildren.addAll(children);
  } else {
    augmentedChildren.add(children);
  }

  return div({}, augmentedChildren);
}
```

Ideally, we'd want to allow for nested children (i.e., fix https://github.com/cleandart/react-dart/issues/60). In that case, we'd especially need the `React.Children` utilities to be able to deal with nested children. For example:
```dart
// Assuming props['children'] can be a single child or a List,
// and assuming children can be nested in Lists: https://github.com/cleandart/react-dart/issues/60
render() {
  return div({},
      react.span({}, 'Sibling to children'),
      props['children']
  );
}

// Workaround until #60 is fixed
render() {
  return div({},
      react.span({}, 'Sibling to children'),
      props['children'] is List ? new JsArray(props['children']) : props['children']
  );
}
```